### ISSUE TITLE
How to track upload progress?

### ISSUE THREAD

**Issue by jaweii** [parse-community/Parse-SDK-Flutter#366]

**Comment by pastordee** [comment]
This would be nice, any word on this.

**Comment by maaeps** [comment]
Currently the [http](https://pub.dev/packages/http) package is used for server communication and as far as I know, there is now way to track the uploading progress using this library.

In order to track the upload of files, we might have to switch to [dio](https://pub.dev/packages/dio).

@phillwiggins Do you think we should do this? As the whole SDK uses the `ParseHTTPClient` (I think), this might actually be possible without to much changes.

**Comment by phillwiggins** [comment]
It does sound like a large change, however, I think you might be right. Except for websockets, I think the entire project depends on ParseHTTPClient which could be swapped out? It's definitely worth a try if you think that's the best option?

**Comment by maaeps** [comment]
I have created a [branch](https://github.com/maaeps/Parse-SDK-Flutter/tree/dio) that uses dio, currently it seems to work, but I will do further testing.

**Comment by maaeps** [comment]
When using it with a heavy app, web freezes sometimes. I think it is related to [this](https://github.com/flutterchina/dio/issues/666).

**Comment by FNPCMDs** [comment]
+1

**Comment by FNPCMDs** [comment]
> I have created a [branch](https://github.com/maaeps/Parse-SDK-Flutter/tree/dio) that uses dio, currently it seems to work, but I will do further testing.

Hi!
Any plans to do a PR with dio package? It may be so useful in android/ios.
Greetings!

**Comment by maaeps** [comment]
The problem is, that at least my test app freezes, using the branch above.
@phillwiggins Do you think It would be an idea to leave everything as it is and just add dio for uploading a file (provinging an additional funktion like `uploadWithProgress` that uses the dio package)?
An other idea would be to create a `parse_server_sdk_extension` package, that adds some extra functionality, only some users need. (Personaly I am already using kind of such an approach for my application.)

**Comment by maaeps** [comment]
I have created a [branch](https://github.com/maaeps/Parse-SDK-Flutter/tree/with-progress) that adds the option to get the upload progress of a file.

**This is only tested on Android.** Further testing is appreciated.
Example:
```dart
file.upload(progressCallback: (int count, int total) => print("$count of $total"));
```
I am still not sure if I will create a PR with this change, as I think, using two network libraries is not ideal.

**Comment by maaeps** [comment]
> When using it with a heavy app, web freezes sometimes. I think it is related to [this](https://github.com/flutterchina/dio/issues/666).

I tried to reproduce this freezes and where not able to do so.
Maybe it has been a bug in flutters beta channel.

**Comment by maaeps** [comment]
I have added the `progressCallback` to the [dio branch](https://github.com/maaeps/Parse-SDK-Flutter/tree/dio).
Testing is highly appreciated. Pleas notify me, if you encounter any ui freezes or other problems.

If we would change the whole SDK to use dio (PR this branch) we could even add send and receive progress indications to the `ParseObject().save()` and `QueryBuilder().query()`. In normal use cases, this might not be used that often, but it might be handy, if many objects are fetched in a single query.

**Comment by FNPCMDs** [comment]
> I have added the `progressCallback` to the [dio branch](https://github.com/maaeps/Parse-SDK-Flutter/tree/dio).
> Testing is highly appreciated. Pleas notify me, if you encounter any ui freezes or other problems.
> 
> If we would change the whole SDK to use dio (PR this branch) we could even add send and receive progress indications to the `ParseObject().save()` and `QueryBuilder().query()`. In normal use cases, this might not be used that often, but it might be handy, if many objects are fetched in a single query.

Thanks u!! I'm going to try it on my app.
Greetings!

**Comment by fischerscode** [comment]
@FNPCMDs 
Are you still using this branch?
Did it work fine?

I am considering to create a similar branch based on 1.0.28.

**Comment by FNPCMDs** [comment]
> @FNPCMDs
> Are you still using this branch?
> Did it work fine?
> 
> I am considering to create a similar branch based on 1.0.28.

Right now I don't have any issues with that branch. I'm using it to develop my android app and everything works like a charm. I think it could be a nice feature in 1.0.28

**Comment by fischerscode** [comment]
@phillwiggins 
What do you think about switching to dio?

**Comment by phillwiggins** [comment]
Risky, but could work quite well. It's a very powerful networking library that could offer some great benefits. I doubt very much that people are using the http library directly, and as it's abstracted by the library we have wrote, there probably wouldn't be many breaking changes?

**Comment by fischerscode** [comment]
Changing http to dio was possible with not to many [changes](https://github.com/parse-community/Parse-SDK-Flutter/compare/release/1.0.27...fischerscode:dio).
As far as I know, there are no breaking changes. Maybe if you provide your own `ParseHTTPClient` but this should be a pretty uncommon case.

I suggest that I will create a branch based on 1.0.28, change everything to dio, implement features like progress callbacks and then test this in my current project. (Maybe @FNPCMDs  can test it, too.)
Then we can consider merging it into 1.0.28.

**Comment by fischerscode** [comment]
I have created a [branch](https://github.com/fischerscode/Parse-SDK-Flutter/tree/dio-1.0.28) based on 1.0.28 that uses dio.
I had no time testing it, yet.
[Migration guide](https://github.com/fischerscode/Parse-SDK-Flutter/blob/create-migrate-guide/docs/migrate-1-0-28.md)

**Comment by FNPCMDs** [comment]
> Changing http to dio was possible with not to many [changes](https://github.com/parse-community/Parse-SDK-Flutter/compare/release/1.0.27...fischerscode:dio).
> As far as I know, there are no breaking changes. Maybe if you provide your own `ParseHTTPClient` but this should be a pretty uncommon case.
> 
> I suggest that I will create a branch based on 1.0.28, change everything to dio, implement features like progress callbacks and then test this in my current project. (Maybe @FNPCMDs can test it, too.)
> Then we can consider merging it into 1.0.28.

Sorry!! I'm getting any notification about this issue. I'm going to test your new branch in the following days. If I find any issues I will let you know. I will gladly help in this SDK.

Greetings!

**Comment by fischerscode** [comment]
No problem :-)
I am currently testing using this [branch](https://github.com/fischerscode/Parse-SDK-Flutter/tree/testing-branch-1). It uses dio and is up to date with 1.0.28. I think dio-1.0.28 is already some commits from 1.0.28 behind. (For example #446)

**Comment by fischerscode** [comment]
@FNPCMDs 
I have just discovered a problem with uploading files. I will try to fix it as soon as possible.

**Comment by fischerscode** [comment]
> @FNPCMDs
> I have just discovered a problem with uploading files. I will try to fix it as soon as possible.

I think it is solved now.

**Comment by FNPCMDs** [comment]
> > @FNPCMDs
> > I have just discovered a problem with uploading files. I will try to fix it as soon as possible.
> 
> I think it is solved now.

Wow that was so fast

**Comment by fischerscode** [comment]
Well... With a break it took 4 hours ;-)

**Comment by fischerscode** [comment]
I am closing this now, as this feature was merged with #459.
